### ISSUE TITLE
Support WebExtensions language packs

### ISSUE THREAD

**Issue by jvillalobos** [mozilla/addons-server#6297]
@zbraniecki and team are working on an implementation of language packs using a WebExtensions package. [Bug 1395363](https://bugzilla.mozilla.org/show_bug.cgi?id=1395363) is tracking their deployment, and AMO support is one of the blockers.

There are a few things we need to do:

* [x] Detect if the XPI is a language pack and give it the correct type, since it's not an extension. #6350.
* [x] In the linter, validate that the add-on _only_ includes the language pack manifest flags. Language packs shouldn't run any code. mozilla/addons-linter#1526.
* [x] We'll probably need to _require_ the use of `strict_min_version` and `strict_max_version`, since legacy language packs have implicit strict compatibility. Also mozilla/addons-linter#1526.
* [x] Look into the legacy validator for any other checks we perform specifically on language packs (I don't remember any), and determine if they should be ported to the linter. Also https://github.com/mozilla/addons-linter/issues/1526.
* [x] Update the language pack import tool (if needed) to support these new packages. #6360

**Comment by zbraniecki** [comment]
> We'll need to know which manifest flags we should look into. @zbraniecki, do you have a spec?

We don't. But we do have a schema and we recognize the langpack by the `manifest.langpack_id` field - https://reviewboard.mozilla.org/r/144184/diff/3

>  We'll probably need to require the use of strict_min_version and strict_max_version, since legacy language packs have implicit strict compatibility.

I have a patch in bug https://bugzilla.mozilla.org/show_bug.cgi?id=1395456 that cleans up setting min/max version for the manifest. Let me know if that seems like the right approach.

Let me know if you need anything else!

**Comment by jvillalobos** [comment]
>  https://reviewboard.mozilla.org/r/144184/diff/3

Okay, I suppose it's best to wait for [bug 1365709](https://bugzilla.mozilla.org/show_bug.cgi?id=1365709) to land so we know what field to look for.

> I have a patch in bug https://bugzilla.mozilla.org/show_bug.cgi?id=1395456 that cleans up setting min/max version for the manifest.

Should be good for our language packs (which are most), but we still need to validate for the handful of community-maintained language packs (and just to play it safe).

**Comment by wagnerand** [comment]
> Update the language pack import tool (if needed) to support these new packages. /cc @wagnerand

I'd like to take the opportunity and automate this process fully. It's a bit silly that every couple of weeks, a human has to go monitor ftp for a day, wait for the final release to get uploaded and then press a button on AMO.

**Comment by zbraniecki** [comment]
With the mozilla-central (as of 2 minutes ago), you can now build the new langpacks by just doing:

```
WEBEXT_LANGPACKS=1 ./mach langpack-fr
```

This will produce a langpack in your `objdir/dist/platform/xpi/` directory.

**Comment by wagnerand** [comment]
That's nice but it still requires human action. I'd like to eliminate the need of having to rely on a human. Humans forget stuff, are on PTO, or sick etc.

**Comment by zbraniecki** [comment]
Apologies. I was just introducing the command people can use to produce a webext-langpack to test it :)
Of course you're right that automated is the way to go!

**Comment by wagnerand** [comment]
No apologies necessary! It's useful info, thank you!

**Comment by jvillalobos** [comment]
@wagnerand Assuming no changes in the build and import process, is there anything that needs to change on addons-server for the import tool to work with WebExtension language packs?

**Comment by zbraniecki** [comment]
Update on our side, everything landed in the product so you can now both build and load the new type of langpacks into Firefox.

I have a couple small items left, but the only major one I know of is that we want to align langpack IDs to get users updated from old type langpacks to new type langpacks seamlessly.

We'd like to switch to the new type of langpacks early in the 58 cycle on Nightly and let it ride trains. I hope it'll work for the AMO team. If you learn that AMO cannot get it working in time, please, flag me so that I can postpone all items that are blocked on this switch. Thanks.

**Comment by wagnerand** [comment]
@jvillalobos I am not familiar with the process behind the scenes. UI wise, there shouldn't be a difference as long as the new langpack files will remain at the current directory on ftp.

**Comment by jvillalobos** [comment]
Most language packs on AMO are imported from Firefox builds, and only for release, so our hard deadline for this is 58 release. I'd like to aim this for 57 instead (November), so there's enough time to iron out any kinks.

**Comment by zbraniecki** [comment]
Heads up! We're ready to make the switch and unless I hear from you, I'm planning to land the switch in central this week and ride the trains for 58 - https://bugzilla.mozilla.org/show_bug.cgi?id=1395363

**Comment by jvillalobos** [comment]
All dependencies have been resolved, so it sounds like this is done.
